New epoll based polling implementation

sreecha opened the pull request.

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1391 @@
+       bit 0    : 1=Active / 0=Orphaned
+       bits 1-n : refcount
+     Ref/Unref by two to avoid altering the orphaned bit */
+  gpr_atm refst;
```
I guess this is really only needed by the pollset_set implementation now (since orphaning will ultimately do an unref for the polling island ref).

This is probably fine for now, though it might be interesting for pollset_set to become just a pointer to a polling island. Then we could drop refst, and mutating the pollset_set could be merges on polling islands.

sreecha replied:
Added a TODO.

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1391 @@
+  bool orphaned;
+
+  grpc_closure *read_closure;
+  grpc_closure *write_closure;
```
Without a watch list needing to be maintained (like in the poll() implementation), there's a mostly lockfree implementation of the read/write_closure code that's available.

We don't need to do so now, but a TODO at least to switch to it would be worthwhile.

sreecha replied:
Added a TODO.

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1391 @@
+  size_t i;
+
+  for (i = 0; i < pi->fd_cnt; i++) {
+    if (remove_fd_refs) {
```
Consider switching the order to ensure no spurious wakeups?

sreecha replied:
Done.

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1391 @@
+        GPR_TIMER_BEGIN("pollset_kick.broadcast", 0);
+        for (worker = p->root_worker.next; worker != &p->root_worker;
+             worker = worker->next) {
+          pthread_kill(worker->pt_id, SIGUSR1);
```
We'll need to figure out API to set this signal number

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1383 @@
+
+  /* The polling island to which this fd belongs to and the mutex protecting the
+     the field */
+  gpr_mu pi_mu;
```
Are we gaining much having pi_mu separate?

Seems like we could get rid of a lock/unlock pair on the pollset_work critical path by just using mu.

sreecha replied:
Probably not much with pi_mu. I will do this in the next PR. Added a TODO for now.

sreecha replied:
removed pi_mu

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1383 @@
+    .shutdown_engine = shutdown_engine,
+};
+
+const grpc_event_engine_vtable *grpc_init_epoll_linux(void) {
```
Probably worth checking whether epoll_create1(...) can succeed at this point in the code, and returning NULL otherwise.

It'll extend our platform reach back a little ways by allowing us to fall back on poll.

ctiller commented on src/core/lib/iomgr/ev_posix.c:
```diff
@@ -62,7 +63,9 @@ typedef struct {
 static const event_engine_factory g_factories[] = {
-    {"poll", grpc_init_poll_posix}, {"legacy", grpc_init_poll_and_epoll_posix},
+    {"poll", grpc_init_poll_posix},
+    {"epoll", grpc_init_epoll_linux},
```
epoll wants to be element 0 in this array (it should be preferred if it's available)

sreecha replied:
Done

ctiller commented on src/core/lib/iomgr/ev_posix.c:
```diff
@@ -163,11 +166,6 @@ void grpc_fd_notify_on_write(grpc_exec_ctx *exec_ctx, grpc_fd *fd,
   g_event_engine->fd_notify_on_write(exec_ctx, fd, closure);
 }
 
-grpc_pollset *grpc_fd_get_read_notifier_pollset(grpc_exec_ctx *exec_ctx,
```
Let's wait on removing this: I want to do some timings with it once the SO_REUSEPORT stuff gets in.

sreecha replied:
Added this back

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1383 @@
+    GPR_TIMER_MARK("pollset_work.kicked_without_pollers", 0);
+    pollset->kicked_without_pollers = 0;
+  } else if (!pollset->shutting_down) {
+    sigemptyset(&new_mask);
```
Having a thread local here tracking the current pollset & worker, and using them in pollset_kick provides a huge performance win. It's probably worth adding that back in here (or results will likely be disappointing).

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1383 @@
+  /* Points to the polling_island this merged into.
+   * If merged_to is not NULL, all the remaining fields (except mu and ref_cnt)
+   * are invalid and must be ignored */
+  struct polling_island *merged_to;
```
Something to consider (but not necessarily change just yet):

If we make ref_cnt and merged_to gpr_atm (merged_to is plausible to do so since it's only set once), then we can do the pointer chase for updates using only atomics, drastically reducing the number of mutexes we need to acquire, and really simplifying the code we need when doing two pointer chases.

We'd likely need to keep epoll_fd & fds around until ref_cnt drops back to zero (should be ok).

It's one place where using atomics might actually result in vastly simplified code.

sreecha replied:
Added a TODO.

ctiller commented:
@dklempner - adding you to the review list because I think you'll have some worthwhile insights.

ctiller commented:
I think I'm done with the initial review. It's looking really good overall.
Many of my comments are likely able to be followed up in a secondary pull request.

I'd like to see some kind of throughput comparison (probably running the c++ qps_unconstrained scenario locally would be enough) between this poller and the poll() based implementation, mostly to guide us in just how important or not some of the optimizations I've suggested are.

sreecha commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1383 @@
+  gpr_mu_unlock(&pollset->mu);
+
+  do {
+    ep_rv = epoll_pwait(epoll_fd, ep_ev, GRPC_EPOLL_MAX_EVENTS, timeout_ms,
```
Fixed the race.

sreecha commented:
I will do a refactor to change ref_cnt and merged_to to atomics. The more I look at this PR code, the more I feel it is confusing and would be very hard to maintain. 

I will do the refactor and update..

sreecha commented:
Addressed all the review comments and the tests are passing. I am going to make one round of refactoring (to simplify the code). Will send it by tomorrow.

sreecha commented:
@ctiller and @dklempner : I have done the following changes: 
- Significantly revised the ref-counting logic (separated locking and ref-counting in to different functions) 
- Removed(`pollset->pi_mu`) lock which was redundant
- Converted `merged_to` and `ref_count` fields in the polling island to atomics. This *significantly* reduced the complexity of `polling_island_lock_pair` function
- Removed taking any locks in `pollset_work_and_unlock` function which is the fast-path

IMO, the code looks much cleaner now and all the tests seem to be passing (the failures are all known except for some error that [generate_projects.sh](https://grpc-testing.appspot.com/job/gRPC_pull_requests/10156/config=opt,language=sanity,platform=linux/testReport/junit/(root)/tests/tools_buildgen_generate_projects_sh/) is reporting)

Can you PTAL ?

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+     Once the ref count becomes zero, this structure is destroyed which means
+     we should ensure that there is never a scenario where a PI_ADD_REF() is
+     racing with a PI_UNREF() that just made the ref_count zero. */
+  gpr_atm ref_count;
```
Since this is a simple ref count (just incs and decs) - we should probably use the gpr_refcount type here and save some LOC.

sreecha replied:
Done. Changed the type to `gpr_refcount`

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+    if (err < 0) {
+      if (errno != EEXIST) {
+        /* TODO: sreek - We need a better way to bubble up this error instead of
+           just logging a message */
```
Right... we should probably be creating an overall grpc_error for this function and returning it up the call stack.

sreecha replied:
Done

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+  err = epoll_ctl(pi->epoll_fd, EPOLL_CTL_ADD,
+                  GRPC_WAKEUP_FD_GET_READ_FD(wakeup_fd), &ev);
+  if (err < 0) {
+    gpr_log(GPR_ERROR,
```
Again, this should be creating a grpc_error and returning it up the stack.

sreecha replied:
Done

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+    if (err < 0 && errno != ENOENT) {
+      /* TODO: sreek - We need a better way to bubble up this error instead of
+      * just logging a message */
+      gpr_log(GPR_ERROR,
```
Create a grpc_error... yada yada

sreecha replied:
Done

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+  if (!is_fd_closed) {
+    err = epoll_ctl(pi->epoll_fd, EPOLL_CTL_DEL, fd->fd, NULL);
+    if (err < 0 && errno != ENOENT) {
+      gpr_log(GPR_ERROR, "epoll_ctl deleting fd: %d failed with error; %s",
```
And here.

sreecha replied:
Done

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+
+  pi->epoll_fd = epoll_create1(EPOLL_CLOEXEC);
+
+  if (pi->epoll_fd < 0) {
```
Create a grpc_error here too.

sreecha replied:
Done

ctiller commented on src/core/lib/iomgr/ev_epoll_linux.c:
```diff
@@ -0,0 +1,1755 @@
+     // polling_island_lock_pair() was called and if so, release the lock only
+     // once. Note: Even if p1 != p2 beforec calling polling_island_lock_pair(),
+     // they might be after the function returns:
+     if (p1 == p2) {
```
Suggest a polling_island_unlock_pair to help with this

sreecha replied:
Great suggestion. Done!

sreecha commented:
@ctiller , I addressed all your review comments. PTAL

ctiller commented:
LGTM. Test failures are unrelated.
